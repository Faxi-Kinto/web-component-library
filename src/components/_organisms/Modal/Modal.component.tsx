import React, { useState, useEffect, useRef, useCallback } from 'react';
import * as Styled from './Modal.styles';
import ReactDOM from 'react-dom';
import classNames from 'classnames';

export type ModalProps = {
  header?: React.ReactNode;
  body?: React.ReactNode;
  footer?: React.ReactNode;
  toggled: boolean;
  className?: string;
  contentClassName?: string;
  parent?: HTMLElement;
  onClickOutOfModal?: () => void;
  headerClassName?: string;
  bodyClassName?: string;
  footerClassName?: string;
  position?: 'top' | 'center' | 'bottom';
  isBanner?: boolean;
  id?: string;
};

const Modal: React.FC<ModalProps> = (props: ModalProps): JSX.Element => {
  const {
    header,
    body,
    footer,
    toggled,
    className,
    contentClassName,
    parent,
    headerClassName,
    bodyClassName,
    footerClassName,
    onClickOutOfModal,
    position = 'center',
    isBanner = false,
    id,
  } = props;
  const myRef = useRef<HTMLDivElement>(null);

  const [isShown, setIsShown] = useState(toggled);

  const handleClickOutside = useCallback(
    (e: MouseEvent): void => {
      if (!isShown) return;
      if (!myRef.current?.contains(e.target as Node)) {
        onClickOutOfModal && onClickOutOfModal();
      }
    },
    [isShown, onClickOutOfModal]
  );

  useEffect(() => {
    setIsShown(toggled);
    if (isShown && !isBanner) {
      document.body.style.overflow = 'hidden';
    }
    if (!isBanner) {
      document.addEventListener('mousedown', handleClickOutside);
      return (): void => {
        document.removeEventListener('mousedown', handleClickOutside);
        document.body.style.overflow = '';
      };
    }
  }, [toggled, isShown, onClickOutOfModal, isBanner, handleClickOutside]);

  return isShown ? (
    ReactDOM.createPortal(
      <Styled.ModalStyles
        id={id}
        className={classNames([
          'wcl-modal',
          { 'wcl-modal--banner': isBanner },
          { 'wcl-modal--top': position === 'top' },
          { 'wcl-modal--center': position === 'center' },
          { 'wcl-modal--bottom': position === 'bottom' },
          className,
        ])}
      >
        <div
          className={classNames(['wcl-modal__content', contentClassName])}
          ref={myRef}
        >
          {header && <div className={headerClassName}>{header}</div>}
          {body && <div className={bodyClassName}>{body}</div>}
          {footer && <div className={footerClassName}>{footer}</div>}
        </div>
      </Styled.ModalStyles>,
      parent ? parent : document.body
    )
  ) : (
    <></>
  );
};

export default Modal;
