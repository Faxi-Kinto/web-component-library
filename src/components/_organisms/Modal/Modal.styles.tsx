import styled from 'styled-components';
import { positionAbsolute, size, flexColumn } from '@faxi/web-css-utilities';
import { modalZIndex } from '@faxi/web-css-utilities/constants';

export const ModalStyles = styled.div`
  ${positionAbsolute('0', '', '', '0')};
  z-index: ${modalZIndex};
  ${size('100%')};
  background-color: #00000050;
  display: flex;
  flex-direction: column;
  align-items: center;

  &.wcl-modal {
    &--banner {
      background: unset;
      height: fit-content;

      &.wcl-modal--top {
        /* unnecessary, defined by default */
      }

      &.wcl-modal--center {
        top: 50%;
      }

      &.wcl-modal--bottom {
        ${positionAbsolute('unset', '', '0', '0')};
      }
    }

    &--top {
      justify-content: flex-start;
    }

    &--center {
      justify-content: center;
    }

    &--bottom {
      justify-content: flex-end;
    }
  }

  .wcl-modal {
    &__content {
      position: relative;
      ${flexColumn()};
      background-color: white;
    }
  }
`;
