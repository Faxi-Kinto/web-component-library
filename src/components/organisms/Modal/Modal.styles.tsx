import styled from 'styled-components';
import { theme } from '../../../config';
import {
  size,
  positionAbsolute,
  padding,
  pxToRem,
  fontSize,
} from '../../../styles/basics';
import { zindexModal } from '../../../styles/constants';
import { flexRow, flexColumn } from '../../../styles/flex';

export const Container = styled.div`
  ${positionAbsolute('0', '', '', '0')};
  z-index: ${zindexModal};
  width: 100vw;
  height: 100vh;
  /* In case we have nested modals */
  & + div {
    opacity: 1;
  }
  .modal-wrapper {
    position: fixed;
    top: 0;
    left: 0;
    z-index: 1001;
    width: 100%;
    height: 100%;
    overflow-x: hidden;
    overflow-y: auto;
    outline: 0;
    background-color: ${theme.pallet.GRAY_SHADOW};

    &__content {
      position: relative;
      background-color: ${theme.pallet.WHITE};
      ${size('725px', '310px')};
      top: 50%;
      left: 50%;
      transform: translate(-50%, -50%);
      ${flexColumn()};

      &__header {
        height: ${pxToRem('55px')};
        background-color: ${theme.pallet.LIGHT_BLUE};
        font-size: ${pxToRem('20px')};
        color: white;
        padding-left: ${pxToRem('45px')};
        ${flexRow('initial', 'center')};
      }

      &__body {
        flex-grow: 1;
        text-align: left;
        ${padding(pxToRem('27px'), pxToRem('60px'), '0', pxToRem('45px'))};
        ${fontSize('16px', '20px')};
        font-weight: lighter;
        letter-spacing: 0;
        color: ${theme.pallet.GRAY_BLUE_2};
      }

      &__line {
        background-color: ${theme.pallet.GRAY_BLUE_2};
        height: ${pxToRem('0.25px')};
        margin: 0;
        width: 100%;
        border-width: 0.02em;
        border-right: 0;
      }

      &__footer {
        height: ${pxToRem('102px')};
        ${flexRow('flex-end', 'center')};
        padding-right: ${pxToRem('35px')};
        button {
          width: ${pxToRem('205px')};
        }
      }
    }
  }
`;
