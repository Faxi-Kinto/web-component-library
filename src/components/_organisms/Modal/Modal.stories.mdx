import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { action } from '@storybook/addon-actions';
import { useState } from 'react';

import Modal from "./Modal.component.tsx";
import Button from "../../_atoms/Button/Button.component.tsx"

<Meta title="Core|Organisms/Modal" />

# Modal Component

This is an implementation of the Modal Component

## Implementation

#### Default

<Preview>
  <Story name="default">
    {() => {
      const [isShowing, setIsShowing] = useState(false)
      return (
      <div>
       <Button onClick={()=>{setIsShowing(!isShowing); action('modal-open') }}>Open Modal</Button>
       <Modal
        toggled={isShowing}  
        header={<span>Confirmation</span>}
        body={<div style={{height:"200px", width:"400px", overflow:"auto"}}>
                <span>"Your community has been successfully registered. A verification link has been sent to your email account. Please verify your email in order to access the created community.Your community has been successfully registered. A verification link has been sent to your email account. Please verify your email in order to access the created community.Your community has been successfully registered. A verification link has been sent to your email account. Please verify your email in order to access the created community.Your community has been successfully registered. A verification link has been sent to your email account. Please verify your email in order to access the created community.Your community has been successfully registered. A verification link has been sent to your email account. Please verify your email in order to access the created community.Your community has been successfully registered. A verification link has been sent to your email account. Please verify your email in order to access the created community.Your community has been successfully registered. A verification link has been sent to your email account. Please verify your email in order to access the created community."</span>   
             </div>}
        footer={<Button type="submit" onClick={()=>{isShowing===true ? setIsShowing(false) : null}}>Ok</Button>}
        onClickOutOfModal={()=>{setIsShowing(false)}}
        position="top"
        customKeyboardEvents={
          [
            {
              keyCodes: ['ShiftLeft', 'KeyA'],
              callbackFn: () => {
                setIsShowing(false);
              },
            },
            {
              keyCodes: ['Escape'],
              callbackFn: () => {
                setIsShowing(false);
              },
            },
          ]
        }
        />
    </div>)
    }}
  </Story>
</Preview>

# Props

<Props of={Modal} />

## Adding a new Modal
- The rendering of Modal component is resposibility of it's parent
- In parent component You must add state variable for that purpose and access to Modal via toggled prop