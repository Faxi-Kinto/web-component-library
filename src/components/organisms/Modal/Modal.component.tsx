import React, { useState, useEffect, useRef } from 'react';
import * as Styled from './Modal.styles';
import ReactDOM from 'react-dom';

export type ModalProps = {
  title?: string;
  body?: string;
  toggled: boolean;
  footer?: JSX.Element;
  onClickOutOfModal?: () => void;
};

const Modal: React.FC<ModalProps> = (props: ModalProps): JSX.Element => {
  const { title, toggled, onClickOutOfModal, body, footer } = props;
  const myRef = useRef<HTMLDivElement>(null);

  const [isShown, setIsShown] = useState(toggled);

  useEffect(() => {
    const handleClickOutside = (e: MouseEvent): void => {
      if (!isShown) return;
      if (!myRef.current?.contains(e.target as Node)) {
        onClickOutOfModal && onClickOutOfModal();
      }
    };
    setIsShown(toggled);
    if (isShown) {
      document.body.style.overflow = 'hidden';
    }
    document.addEventListener('mousedown', handleClickOutside);
    return (): void => {
      document.removeEventListener('mousedown', handleClickOutside);
      document.body.style.overflow = 'unset';
    };
  }, [toggled, isShown, onClickOutOfModal]);

  return isShown ? (
    ReactDOM.createPortal(
      <Styled.Container>
        <div className="modal-wrapper">
          <div className="modal-wrapper__content" ref={myRef}>
            <div className="modal-wrapper__content__header">{title}</div>
            <div className="modal-wrapper__content__body">{body}</div>
            <hr className="modal-wrapper__content__line" />
            <div className="modal-wrapper__content__footer">{footer}</div>
          </div>
        </div>
      </Styled.Container>,
      document.body
    )
  ) : (
    <></>
  );
};

export default Modal;
