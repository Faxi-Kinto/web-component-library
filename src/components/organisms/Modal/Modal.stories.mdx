import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { action } from '@storybook/addon-actions';
import { useState } from 'react';

import Modal from "./Modal.component.tsx";
import Button from "../../atoms/Button/Button.component.tsx"

<Meta title="Core|Organisms/Modal" />

# Modal Component

This is an implementation of the Modal Component

## Implementation

#### Default

<Preview>
  <Story name="default">
    {() => {
      const [isShowing, setIsShowing] = useState(false)
      return (
      <div>
       <Button onClick={()=>{setIsShowing(true); action('modal-open') }}>Open Modal</Button>
       <Modal
        toggled={isShowing}  
        title={<span>Confirmation</span>}
        body={<span>"Your community has been successfully registered. A verification link has been sent to your email account. Please verify your email in order to access the created community."</span>}
        footer={<Button type="submit" onClick={()=>{isShowing===true ? setIsShowing(false) : null}}>Ok</Button>}
        onClickOutOfModal={()=>{setIsShowing(false)}}
        />
    </div>)
    }}
  </Story>
</Preview>

# Props

<Props of={Modal} />

## Adding a new Modal
- The rendering of Modal component is resposibility of it's parent
- In parent component You must add state variable for that purpose and access to Modal via toggled prop