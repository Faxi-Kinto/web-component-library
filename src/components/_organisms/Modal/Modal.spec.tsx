import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import Modal from './Modal.component';

test('Modal renders correctly', () => {
  const renderer = ShallowRenderer.createRenderer();

  const component = renderer.render(
    <Modal
      toggled={true}
      header={<span>Test</span>}
      body={<span>test body</span>}
      footer={<button>test</button>}
    />
  );

  expect(component).toMatchSnapshot();
});
