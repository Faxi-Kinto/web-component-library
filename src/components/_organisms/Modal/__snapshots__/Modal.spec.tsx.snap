// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
Object {
  "$$typeof": Symbol(react.portal),
  "children": <ForwardRef(styled.div)
    className="wcl-modal wcl-modal--center"
  >
    <div
      className="wcl-modal__content"
    >
      <div>
        <span>
          Test
        </span>
      </div>
      <div>
        <span>
          test body
        </span>
      </div>
      <div>
        <button>
          test
        </button>
      </div>
    </div>
  </ForwardRef(styled.div)>,
  "containerInfo": <body />,
  "implementation": null,
  "key": null,
}
`;
